--- gdb-7.2/gdb/testsuite/gdb.base/break-interp.exp	2012-01-14 21:55:34.500450965 +0100
+++ gdb-7.2-x/gdb/testsuite/gdb.base/break-interp.exp	2012-01-14 21:53:01.557380926 +0100
@@ -109,9 +109,10 @@ proc strip_debug {dest} {
 }
 
 # Former symbol for solib changes notifications was _dl_debug_state, newer one
-# is _dl_debug_notify, the right one one traps by `set stop-on-solib-events 1'.
+# is dl_main (in fact _dl_debug_notify but it is inlined without any extra
+# debug info), the right one one traps by `set stop-on-solib-events 1'.
 
-set solib_bp {(_dl_debug_state|_dl_debug_notify)}
+set solib_bp {(_dl_debug_state|dl_main)}
 
 # Implementation of reach.
 
