http://sourceware.org/ml/gdb-cvs/2012-01/msg00205.html

Index: gdb-7.2/gdb/gdb_vecs.h
===================================================================
--- /dev/null
+++ gdb-7.2/gdb/gdb_vecs.h
@@ -0,0 +1,28 @@
+/* Some commonly-used VEC types.
+
+   Copyright (C) 2012 Free Software Foundation, Inc.
+
+   This file is part of GDB.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
+
+
+#ifndef GDB_VECS_H
+#define GDB_VECS_H
+
+#include "vec.h"
+
+DEF_VEC_P (char_ptr);
+
+#endif /* GDB_VECS_H */
Index: gdb-7.2/gdb/ada-lang.c
===================================================================
--- gdb-7.2.orig/gdb/ada-lang.c
+++ gdb-7.2/gdb/ada-lang.c
@@ -57,6 +57,7 @@
 #include "observer.h"
 #include "vec.h"
 #include "stack.h"
+#include "gdb_vecs.h"
 
 #include "psymtab.h"
 
@@ -5242,8 +5243,6 @@ symbol_completion_match (const char *sym
   return sym_name;
 }
 
-DEF_VEC_P (char_ptr);
-
 /* A companion function to ada_make_symbol_completion_list().
    Check if SYM_NAME represents a symbol which name would be suitable
    to complete TEXT (TEXT_LEN is the length of TEXT), in which case
Index: gdb-7.2/gdb/charset.c
===================================================================
--- gdb-7.2.orig/gdb/charset.c
+++ gdb-7.2/gdb/charset.c
@@ -28,6 +28,7 @@
 #include "vec.h"
 #include "environ.h"
 #include "arch-utils.h"
+#include "gdb_vecs.h"
 
 #include <stddef.h>
 #include "gdb_string.h"
@@ -703,8 +704,6 @@ wchar_iterate (struct wchar_iterator *it
 
 extern initialize_file_ftype _initialize_charset; /* -Wmissing-prototype */
 
-DEF_VEC_P (char_ptr);
-
 static VEC (char_ptr) *charsets;
 
 #ifdef PHONY_ICONV
Index: gdb-7.2/gdb/tracepoint.h
===================================================================
--- gdb-7.2.orig/gdb/tracepoint.h
+++ gdb-7.2/gdb/tracepoint.h
@@ -22,6 +22,7 @@
 
 #include "breakpoint.h"
 #include "target.h"
+#include "gdb_vecs.h"
 
 /* A trace state variable is a value managed by a target being
    traced. A trace state variable (or tsv for short) can be accessed
@@ -124,8 +125,6 @@ extern char *default_collect;
 
 /* Struct to collect random info about tracepoints on the target.  */
 
-DEF_VEC_P (char_ptr);
-
 struct uploaded_tp
 {
   int number;
